src/parser/regexp.bnf

```diff
@@ -35,96 +37,112 @@ GROUP_NAME              ([\w$]|\\'u'[0-9a-fA-F]{4}|\\'u{'[0-9a-fA-F]{1,}'}')+
 
-{ESC}'/'                 return 'ESC_CHAR'
+{ESC}[tnrdDsSwWvf]                  return 'META_CHAR'
 
-{ESC}{CHAR}              return 'ESC_CHAR'
-{ESC}[*?+\[()]           return 'ESC_CHAR'
+{ESC}'/'                            return 'ESC_CHAR'
 
-<class>'('               return 'CHAR'
-<class>')'               return 'CHAR'
+{ESC}{CHAR}                         return 'ESC_CHAR'
+{ESC}[*?+\[()]                      return 'ESC_CHAR'
 
-'(?='                    return 'POS_LA_ASSERT'
-'(?!'                    return 'NEG_LA_ASSERT'
+<class,u_class>'('                  return 'CHAR'
+<class,u_class>')'                  return 'CHAR'
 
-'(?<='                   return 'POS_LB_ASSERT'
-'(?<!'                   return 'NEG_LB_ASSERT'
+'(?='                               return 'POS_LA_ASSERT'
+'(?!'                               return 'NEG_LA_ASSERT'
 
-'(?:'                    return 'NON_CAPTURE_GROUP'
+'(?<='                              return 'POS_LB_ASSERT'
+'(?<!'                              return 'NEG_LB_ASSERT'
 
-'(?<'{GROUP_NAME}'>'     { yytext = yytext.slice(3, -1); return 'NAMED_CAPTURE_GROUP' }
+'(?:'                               return 'NON_CAPTURE_GROUP'
 
-'('                      return 'L_PAREN'
-')'                      return 'R_PAREN'
+'(?<'{GROUP_NAME}'>'                { yytext = yytext.slice(3, -1); return 'NAMED_CAPTURE_GROUP' }
 
-<class>[*?+[^$]          return 'CHAR'
-<class>{ESC}']'          return 'ESC_CHAR'
+'('                                 return 'L_PAREN'
+')'                                 return 'R_PAREN'
 
-<class>']'               { this.popState(); return 'R_BRACKET' }
+<class,u_class>[*?+[^$]             return 'CHAR'
+<class,u_class>{ESC}']'             return 'ESC_CHAR'
 
-'^'                      return 'BOS'
-'$'                      return 'EOS'
+<class,u_class>']'                  { this.popState(); return 'R_BRACKET' }
 
-'*'                      return 'STAR'
-'?'                      return 'Q_MARK'
-'+'                      return 'PLUS'
+'^'                                 return 'BOS'
+'$'                                 return 'EOS'
 
-'|'                      return 'BAR'
+'*'                                 return 'STAR'
+'?'                                 return 'Q_MARK'
+'+'                                 return 'PLUS'
 
-'.'                      return 'ANY'
+'|'                                 return 'BAR'
 
-'/'                      return 'SLASH'
+'.'                                 return 'ANY'
 
-{CHAR}                   return 'CHAR'
+'/'                                 return 'SLASH'
 
-'[^'                     { this.pushState('class'); return 'NEG_CLASS' }
-'['                      { this.pushState('class'); return 'L_BRACKET' }
+{CHAR}                              return 'CHAR'
+
+'[^'                                { this.pushState(['u','xu'].includes(this.getCurrentState()) ? 'u_class' : 'class'); return 'NEG_CLASS' }
```

Comment:
Let's not allocate an array each time we entry a class. We can do simple `if` check, or allocate it once at the top as a constant, and use here. You can also create a helper function, and reuse it below as well.

Reply:
Ok, changed it.